pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapFillGroupByDataTableReducer.java

```diff
@@ -0,0 +1,725 @@
+          row[j] = columnDataTypes[j].convertAndFormat(row[j]);
+        }
+        rows.add(row);
+      }
+      brokerResponseNative.setResultTable(new ResultTable(prePostAggregationDataSchema, rows));
+    }
+  }
+
+  List<Object[]> gapFill(List<Object[]> resultRows, ColumnDataType[] resultColumnDataTypes) {
+    int limit = _queryContext.getLimit();
+    int numResultColumns = resultColumnDataTypes.length;
+    List<Object[]> gapfillResultRows = new ArrayList<>(limit);
+    long step = _dateTimeGranularity.granularityToMillis();
+    int index = 0;
+    for (long time = _startMs; time + 2 * step <= _endMs; time += step) {
+      Set<Key> keys = new HashSet<>(_primaryKeys);
+      while (index < resultRows.size()) {
+        long timeCol = _dateTimeFormatter.fromFormatToMillis((String) resultRows.get(index)[0]);
+        if (timeCol < time) {
+          index++;
+        } else if (timeCol == time) {
+          gapfillResultRows.add(resultRows.get(index));
+          if (gapfillResultRows.size() == limit) {
+            return gapfillResultRows;
+          }
+          Key key = constructKey(resultRows.get(index));
+          keys.remove(key);
+          for (int colIndex = 2; colIndex < numResultColumns; colIndex++) {
+            _previous.put(key, resultRows.get(index));
+          }
+          index++;
+        } else {
+          break;
+        }
+      }
+      for (Key key : keys) {
+        Object[] gapfillRow = new Object[numResultColumns];
+        gapfillRow[0] = _dateTimeFormatter.fromMillisToFormat(time);
+        for (int i = 0; i < _numOfKeyColumns; i++) {
+          gapfillRow[i + 1] = key.getValues()[i];
+        }
+
+        for (int i = _numOfKeyColumns + 1; i < numResultColumns; i++) {
+          gapfillRow[i] = getFillValue(i, key, resultColumnDataTypes[i]);
+        }
+        gapfillResultRows.add(gapfillRow);
+        if (gapfillResultRows.size() == limit) {
+          return gapfillResultRows;
+        }
+      }
+    }
+    return gapfillResultRows;
+  }
+
+  Object getFillValue(int columIndex, Object key, ColumnDataType dataType) {
+    ExpressionContext expressionContext = _queryContext.getSelectExpressions().get(columIndex);
+    if (expressionContext.getFunction() != null
+        && expressionContext.getFunction().getFunctionName().equalsIgnoreCase("fill")) {
+      List<ExpressionContext> args = expressionContext.getFunction().getArguments();
+      if (args.get(1).getLiteral() == null) {
+        throw new UnsupportedOperationException("Wrong Sql.");
+      }
+      FillType fillType = FillType.valueOf(args.get(1).getLiteral());
+      if (fillType == FillType.FILL_DEFAULT_VALUE) {
+        // TODO: may fill the default value from sql in the future.
+        return SelectionOperatorUtils.getDefaultValue(dataType);
+      } else if (fillType == FillType.FILL_PREVIOUS_VALUE) {
+        if (_previous.containsKey(key)) {
+          return _previous.get(key)[columIndex];
```

Comment:
For performance reasons it's much better to get and do a null check:

```java
Object[] value = _previous.get(key);
if (value != null) {
  return value[columnIndex];
}
```

This results in doing half the work to locate the value.

Reply:
Great comment! Done